### PR TITLE
Added render tests for code schools

### LINKED ISSUE #390: Create e2e coverage using Cypress for /code_schools
# Feature request
**Is your feature request related to a problem? Please describe.**
The Code Schools page is more complex as it relies upon http requests to work (for the most part).

**Describe the solution you'd like**
Use Cypress to create e2e test coverage thereby making our team more confident in deploys.

**What do these tests look like?**
- [x] Assert that when the user navigates to the page, the code schools are all rendered on the page
- [x] Assert that only `Cincy Code It Bootcamps` and `Tech Elevator` appear after typing `Ohio` into the react-select input.
- [x] Assert that, after clicking the "VA-Approved Schools" button, all rendered code schools accept the GI Bill
- [x] Assert that, after clicking the "Online Schools" button, all rendered code schools have an online program.
- [x] Assert that no code schools render after typing `Alaska` into the react-select input.
- [ ] Assert that no code schools render after typing `Alaska` into the react-select input, and then (after clicking the "All Schools" button) all code schools render.
- [ ] Assert that page fails gracefully when the server is down (this may be an instance where you write a test which fails, and leads to feature work as happened in #431)

### PR DESCRIPTION
**Pull request by phongvdoan**
# Description of changes

Added tests for inputting 'Ohio', clicking on VA Approved Schools, and inputting  'Alaska 
# Issue Resolved

Partial Fix To  #390 

## Screenshots/GIFs

Screen Shot 2019-05-07 at 12 06 43 AM

### CODE REVIEW

**Comment by kylemh** [comment]
🤖 **This is a bot** 🤖<br /><br />🎉 Deployed Storybook preview! 🎉<br /><br />Click the link at the bottom of this comment to see it live.

Built with commit 7d71062ece7c8b7edc0adda362f1384996fa58ab

https://deploy-preview-468--operation-code-storybook.netlify.com

File: cypress/integration/code-schools.spec.js [hunk 1]
```diff
@@ -14,4 +14,31 @@ describe('code schools', function() {
     // 40 is arbitrary, but it proves that the API is working and leading to rendered content
     cy.get('[data-testid="SchoolCard"]').should('have.length.greaterThan', 40);
   });
+
+  it('should render Cincy and Tech Elevator after selecting Ohio', () => {
+    cy.wait('@codeSchools');
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: Seeing it occur in writing makes me realize we could just put `cy.wait('@codeSchools')` at the end of the `beforeEach()`.

File: cypress/integration/code-schools.spec.js [hunk 2]
```diff
@@ -14,4 +14,31 @@ describe('code schools', function() {
...
+    cy.get('input#react-select-state_select-input')
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: Try passing `data-testid` to the react select component. Then you can target a selector that's human-readable.

File: cypress/integration/code-schools.spec.js [hunk 3]
```diff
@@ -14,4 +14,31 @@ describe('code schools', function() {
...
+      .type('Ohio', { force: true })
+      .type('{enter}');
+
+    cy.get('[data-testid="SchoolCard"]').should('have.length', 2);
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: This is good. You're asserting that there only exists 2 SchoolCards, but we should also look more closely at the DOM of each of those cards to ensure they are indeed the two schools your test is named after!

File: cypress/integration/code-schools.spec.js [hunk 4]
```diff
@@ -14,4 +14,31 @@ describe('code schools', function() {
...
+  });
+
+  it('should render GI Bill accepted schools after hitting VA button', () => {
+    cy.wait('@codeSchools');
+    cy.contains('VA Approved Schools').click();
+
+    cy.get('[data-testid="SchoolCard"]').should('have.length.greaterThan', 20);
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: This won't quite be what I'm looking for.

Does your assertion actually guarantee that the >20 SchoolCards you've rendered take the GI Bill? What if, for example, the button broke... You'd just be matching the first test, and that is what's called a "false positive". A test that passes, but shouldn't.

If you look at the source code for the SchoolCard component, you can search for references to the GI Bill to see what sets it apart from other SchoolCards. You could assert that the text `GI Bill` exists, but that would be true for all cards (as it acts as a label here:

```javascript
<div className={styles.cardBlock}>
  <span className={styles.cardBlockTitle}>Accepts GI Bill</span>
  {hasGiBill ? 'Yes' : 'No'}
</div>
```

The ribbon, however, is conditionally rendered based upon whether or not it takes the GI Bill. I recommend passing `data-testid` to the div that wraps the ribbon, and find assert that _every_ rendered SchoolCard has it.

File: cypress/integration/code-schools.spec.js [hunk 5]
```diff
@@ -14,4 +14,31 @@ describe('code schools', function() {
...
+  });
+
+  // /Online Schools test
+
+  it('should render no schools after selecting Alaska', () => {
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: 😂 poor Alaska

File: cypress/integration/code-schools.spec.js (hunk 1 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Thats what I was thinking.

Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Completed this

File: cypress/integration/code-schools.spec.js (hunk 3 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: When you get a chance, can you look at this one. I replaced it with`cy.get('[data-testid="SchoolCard"]')
      .find('img[alt="Cincy Code IT Bootcamps logo"]', 'img[alt="Tech Elevator logo"]')
      .should('be.visible');`

File: cypress/integration/code-schools.spec.js (hunk 4 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Completed with data-testid in GI Bill Ribbon

File: components/Cards/SchoolCard/SchoolCard.js [hunk 6]
```diff
@@ -70,7 +70,11 @@ export default class SchoolCard extends Component {
 
     return (
       <Card className={styles.SchoolCard} hasAnimationOnHover={false} data-testid="SchoolCard">
-        {hasGiBill && <div className={styles.giBillRibbon}>GI Bill</div>}
+        {hasGiBill && (
+          <div className={styles.giBillRibbon} data-testid="GIBill">
```
Comment (kylemh) [inline components/Cards/SchoolCard/SchoolCard.js]: Can we be more descriptive with this? Maybe `data-testid="GI Bill Ribbon"`?

That will make my desired assertion read rather nicely.

File: cypress/integration/code-schools.spec.js [hunk 7]
```diff
@@ -2,16 +2,39 @@ describe('code schools', function() {
...
+
+    cy.get('[data-testid="SchoolCard"]')
+      .find('img[alt="Cincy Code IT Bootcamps logo"]', 'img[alt="Tech Elevator logo"]')
+      .should('be.visible');
+  });
+
+  it('should render GI Bill accepted schools after hitting VA button', () => {
+    cy.contains('VA Approved Schools').click();
+
+    cy.get('[data-testid="GIBill"]').should('have.length.greaterThan', 20);
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: Better, but we can improve even more.

Instead, let's be very crystal clear in our test so every human can understand:

```javascript
cy.get('[data-testid="SchoolCard"]').each(card => {
  cy.wrap(card)
    .find('[data-testid="GI Bill Ribbon"]')
    .should('exist');
});
```

This ensures that each and every rendered SchoolCard is one that accepts the GI Bill.

File: cypress/integration/code-schools.spec.js [hunk 8]
```diff
@@ -2,16 +2,39 @@ describe('code schools', function() {
...
+    cy.get('input#react-select-state_select-input')
+      .type('Ohio', { force: true })
+      .type('{enter}');
+
+    cy.get('[data-testid="SchoolCard"]')
+      .find('img[alt="Cincy Code IT Bootcamps logo"]', 'img[alt="Tech Elevator logo"]')
+      .should('be.visible');
+  });
+
+  it('should render GI Bill accepted schools after hitting VA button', () => {
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: We can be more explicit in our test names. Maybe: `should only render schools that accept the GI Bill after clicking the 'VA Approved Schools' button`

File: cypress/integration/code-schools.spec.js [hunk 9]
```diff
@@ -2,16 +2,39 @@ describe('code schools', function() {
...
     cy.get('[data-testid="SchoolCard"]').should('have.length.greaterThan', 40);
   });
+
+  it('should render Cincy and Tech Elevator after selecting Ohio', () => {
+    cy.get('input#react-select-state_select-input')
+      .type('Ohio', { force: true })
+      .type('{enter}');
+
+    cy.get('[data-testid="SchoolCard"]')
+      .find('img[alt="Cincy Code IT Bootcamps logo"]', 'img[alt="Tech Elevator logo"]')
```
Comment (kylemh) [inline cypress/integration/code-schools.spec.js]: You can be even simpler:

```javascript
cy.get('[data-testid="SchoolCard"]')
    .find('Cincy Code IT Bootcamps', 'Tech Elevator')
    .should('exist');
```

File: cypress/integration/code-schools.spec.js (hunk 2 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Im going to have to work on this one in a bit. I can pass an id, but will need to do a custom.component for a custom selector. I just saw your latest pull request and will shift fire once i hop on my laptop

File: cypress/integration/code-schools.spec.js (hunk 9 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Ah ok. .find for the text.

File: cypress/integration/code-schools.spec.js (hunk 2 above)
Comment (kylemh) [reply cypress/integration/code-schools.spec.js]: I don't think you'll need a custom component. You can simply pass `data-testid` to the `ThemedReactSelect` component on the code schools page

Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: I will try it again. Last night at 2 am i was getting errors left and right for it, maybe i was just too tired.

File: components/Cards/SchoolCard/SchoolCard.js (hunk 6 above)
Comment (phongvdoan) [reply components/Cards/SchoolCard/SchoolCard.js]: Completed this change

Comment (kylemh) [reply components/Cards/SchoolCard/SchoolCard.js]: Did you forget to push, perhaps?

File: cypress/integration/code-schools.spec.js (hunk 2 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Kyle, after doing so, data-testid is not showing up as a selector. Am I doing it wrong?
Screen Shot 2019-05-07 at 11 05 15 AM

Screen Shot 2019-05-07 at 11 08 47 AM

File: components/Cards/SchoolCard/SchoolCard.js (hunk 6 above)
Comment (phongvdoan) [reply components/Cards/SchoolCard/SchoolCard.js]: Sorry doing it now.

Comment (phongvdoan) [reply components/Cards/SchoolCard/SchoolCard.js]: got to fix lint error

File: cypress/integration/code-schools.spec.js (hunk 7 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: changed

File: cypress/integration/code-schools.spec.js (hunk 9 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: completed

File: cypress/integration/code-schools.spec.js (hunk 2 above)
Comment (kylemh) [reply cypress/integration/code-schools.spec.js]: No, you're right.

Looks like [that's unfortunately unexpected](https://github.com/JedWatson/react-select/issues/3149).

Go ahead and continue selecting the component as you were initially.

Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: Yeah that was the one I was looking at this morning. Ok revert back

File: cypress/integration/code-schools.spec.js (hunk 9 above)
Comment (phongvdoan) [reply cypress/integration/code-schools.spec.js]: This threw an error, so I had to use
```JS  
.should('contain', 'Cincy Code IT Bootcamps')
```
which checked to make sure the SchoolCard contained the string.

File: components/Cards/SchoolCard/SchoolCard.js [hunk 10]
```diff
@@ -70,10 +70,20 @@ export default class SchoolCard extends Component {
...
 
         <div className={styles.cardBrand}>
-          <img src={props.logoSource} alt={`${props.name} logo`} height="150" />
+          <img
+            src={props.logoSource}
+            alt={`${props.name} logo`}
+            data-testid="School logo"
+            height="150"
+          />
+          <p>{props.name}</p>
```
Comment (kylemh) [inline components/Cards/SchoolCard/SchoolCard.js]: So, we don't want to change the component so drastically (adding new elements) just for the sake of the test, so you should remove this line; however, your tests have highlighted an issue with our current SchoolCard implementation. Users that navigate the web with screen readers, will be left to assume the name of the Code School via the image's alt tag - that's not ideal.

Let's use the `ScreenReaderOnly` component at the top of the card to declare the name of the code school. Blind users will be happier, and our tests should be very simple.

**Comment by kylemh** [comment]
Gonna merge it and resolve my nitpicks for ya. Very good work here!

**Pull request merged**